kong/init.lua

```diff
@@ -247,6 +234,57 @@ local function flush_delayed_response(ctx)
 end
 
 
+local function parse_declarative_config(config)
+  if not config.database == "off" then
+    return {}
+  end
+
+  if not config.declarative_config then
+    return {}
+  end
+
+  local dc = declarative.init(config)
+  local entities, err = dc:parse_file(config.declarative_config)
+  if not entities then
+    return nil, "error parsing declarative config file " ..
+                config.declarative_config .. ":\n" .. err
+  end
+
+  return entities
+end
+
+
+local function load_declarative_config(config, entities)
+  if not config.database == "off" then
+    return true
+  end
+
+  if not config.declarative_config then
+    -- no configuration yet, just build empty plugins map
+    build_plugins_map(kong.db, utils.uuid())
+    return true
+  end
+
+  local opts = {
+    name = "declarative_config",
+    timeout = 3600,
```

Comment:
We need to specify `exptime` here as well, or else this timeout setting will not take effect. Also, a 1h timeout? Really? I am very surprised.

Reply:
> We need to specify exptime here as well, or else this timeout setting will not take effect. 

The library sets `exptime`.

> Also, a 1h timeout? Really? I am very surprised.

This is initialization code and there's no recovery from a timeout in this situation, so the timeout just needs to be high enough to cover every possible valid scenario, so this is effectively meant "no timeout". The arbitrary default used elsewhere in the code is 1 minute. I can just use that, so at least the user gets an error message sooner if they have the patience to wait while the system fails to start.

Reply:
> The library sets `exptime`.

Oh, your abstraction in `concurrency.lua`, I see :+1: